internal/app/machined/pkg/system/services/etcd.go

```diff
@@ -269,28 +269,31 @@ func addMember(r runtime.Runtime, addrs []string, name string) (*clientv3.Member
 		}
 	}
 
-	add, err := client.MemberAdd(context.Background(), addrs)
+	add, err := client.MemberAdd(ctx, addrs)
 	if err != nil {
 		return nil, 0, err
 	}
 
-	list, err = client.MemberList(context.Background())
+	list, err = client.MemberList(ctx)
 	if err != nil {
 		return nil, 0, err
 	}
 
 	return list, add.Member.ID, nil
 }
 
-func buildInitialCluster(r runtime.Runtime, name, ip string) (initial string, err error) {
+func buildInitialCluster(ctx context.Context, r runtime.Runtime, name, ip string) (initial string, err error) {
 	err = retry.Constant(10*time.Minute, retry.WithUnits(3*time.Second), retry.WithJitter(time.Second)).Retry(func() error {
 		var (
 			peerAddrs = []string{"https://" + net.FormatAddress(ip) + ":2380"}
 			resp      *clientv3.MemberListResponse
 			id        uint64
 		)
 
-		resp, id, err = addMember(r, peerAddrs, name)
+		attemptCtx, attemptCtxCancel := context.WithTimeout(ctx, 30*time.Second)
```

Comment:
added context propagation and limiting each attempt duration to make sure we do more attempts during those 10 minutes